[optimizer] refactor SGD to use functional API

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50630 [dist_optim] add warning to distributed optimizer
* #50620 [dist_optim] add distributed functional AdamW optimizer
* #50619 [dist_optim] add distributed functional RMSprop optimizer
* #50623 [dist_optim] add distributed functional Adadelta optimizer
* #50618 [dist_optim] add distributed functional sgd optimizer
* #50624 [dist_optim] add distributed functional Adam optimizer
* #50411 [optimizer] refactor AdamW to use functional API
* #50410 [optimizer] refactor RMSProp to use functional API
* #50409 [optimizer] refactor Adadelta to use functional API
* **#45597 [optimizer] refactor SGD to use functional API**

Differential Revision: [D25932773](https://our.internmc.facebook.com/intern/diff/D25932773)

wanchaol committed:
[optimizer] refactor SGD to use functional API

[ghstack-poisoned]

wanchaol committed:
Update on "[optimizer] refactor SGD to use functional API"




[ghstack-poisoned]

wanchaol committed:
Update on "[optimizer] refactor SGD to use functional API"




[ghstack-poisoned]

wanchaol committed:
Update on "[optimizer] refactor SGD to use functional API"




[ghstack-poisoned]

wanchaol committed:
Update on "[optimizer] refactor SGD to use functional API"




[ghstack-poisoned]

wanchaol committed:
Update on "[optimizer] refactor SGD to use functional API"




[ghstack-poisoned]

wanchaol committed:
Update on "[optimizer] refactor SGD to use functional API"




[ghstack-poisoned]

wanchaol committed:
Update on "[optimizer] refactor SGD to use functional API"




[ghstack-poisoned]

wanchaol committed:
Update on "[optimizer] refactor SGD to use functional API"


Differential Revision: [D25932773](https://our.internmc.facebook.com/intern/diff/D25932773)

[ghstack-poisoned]

albanD commented on torch/optim/sgd.py:
```diff
@@ -86,29 +87,39 @@ def step(self, closure=None):
-                        buf = param_state['momentum_buffer'] = torch.clone(d_p).detach()
+                if p.grad is not None:
+                    params_with_grad.append(p)
+                    d_p = p.grad
```
nit: you can do directly `d_p_list.append(p.grad)`

albanD commented on torch/optim/functional.py:
```diff
@@ -96,3 +96,39 @@ def adam(params: List[Tensor],
+            buf = momentum_buffer_list[i]
+
+            if buf is None:
+                buf = torch.clone(d_p).detach()
```
nit: why the detach here? Are there any case where this function is called outside of a `torch.no_grad()` block?

wanchaol replied:
not really, it should all under a `torch.no_grad()` block, I just copied this from the sgd.py, if there's no need to detach here, I can remove it.

albanD replied:
Ho I didn't realize it was in the original version. I think you should keep it as is here to reduce any chance to break stuff in this refactor. :)

wanchaol committed:
Update on "[optimizer] refactor SGD to use functional API"


Differential Revision: [D25932773](https://our.internmc.facebook.com/intern/diff/D25932773)

[ghstack-poisoned]

albanD reviewed (approved):
LGTM !

wanchaol committed:
Update on "[optimizer] refactor SGD to use functional API"


Differential Revision: [D25932773](https://our.internmc.facebook.com/intern/diff/D25932773)

[ghstack-poisoned]
